haskell/test/Test/Plist/Property.hs

```diff
@@ -0,0 +1,128 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+module Test.Plist.Property
+  ( hprop_Plist
+  ) where
+
+import Universum hiding (drop, swap)
+
+import qualified Data.List as DL
+import Data.List ((!!))
+import Fmt (build, unlinesF)
+import Hedgehog hiding (assert)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+import Lorentz hiding (assert, and, div, now, (>>))
+import Test.Cleveland
+
+import Ligo.BaseDAO.Types
+import SMT.Model.BaseDAO.Proposal.Plist
+import Test.Ligo.BaseDAO.Plist (genProposalKeyList)
+import Test.Plist.Type
+import Test.Plist.Contract
+
+
+genPlistParameter :: MonadGen m => [ProposalKey] -> [ProposalKey] -> m PlistParameter
+genPlistParameter keyList keyList2 = do
+  (i :: Int) <- Gen.integral (Range.constant 0 (length keyList - 1))
+  let k = keyList !! i
+  (i2 :: Int) <- Gen.integral (Range.constant 0 (length keyList2 - 1))
+  let k2 = keyList2 !! i2
+  Gen.choice
+    [ pure $ Insert k2
+    , pure $ Mem k2
+    , pure $ Mem k
+    , pure $ Delete k
+    , pure $ Pop ()
+    ]
+
+
+genPlistParameters :: MonadGen m => m ([ProposalKey], [PlistParameter])
+genPlistParameters = do
+  initSize <- Gen.integral (Range.constant 1 10)
+
+  -- Init List to be used as storage
+  keyList <- genProposalKeyList initSize
+
+  -- A list to pick new key from
+  keyList2 <- genProposalKeyList 30
+
+  params <- Gen.list (Range.linear 10 20) $ genPlistParameter keyList keyList2
+
+  pure (keyList, params)
+
+
+hprop_Plist :: Property
+hprop_Plist =
+  withTests 5 $ property $ do
+    plistTests
+
+plistTests :: PropertyT IO ()
+plistTests = do
+  (keyList, params) <- forAll genPlistParameters
+  testScenarioProps $
+    scenarioEmulated $ do
+      let defaultStorage = PlistStorage (plistFromList keyList) False
+
+      -- Originate plist contract
+      plistContract <- originateTypedSimple @PlistParameter "PlistContract" defaultStorage plistContractLigo
+
+      callContractLoop params (TAddress $ chAddress plistContract) defaultStorage
+
+callContractLoop :: MonadEmulated caps base m => [PlistParameter] -> TAddress PlistParameter -> PlistStorage -> m ()
+callContractLoop param addr haskellStore =
+  case param of
+    p:ps -> do
+      let newHaskellStore = callHaskell haskellStore p
+      newLigoStore <- callLigo addr p
+
+      assert (newHaskellStore == newLigoStore) $
```

Comment:
Here it seems that we are using the `Eq` instance of `ProposalDoublyLinkedList` for comparision. I think it is safer to convert the `ProposalDoublyLinkedList` to a list and compare it to the result list from Haskell version. So I think both callLigo and callHaskell should return the list along with memresult.

Reply:
Make sense, I've udpated both `callLigo` and `callHaskell` to return `([ProposalKey], Bool)` instead now.